src/lv_misc/lv_color.c

```diff
@@ -105,39 +106,66 @@ lv_color_t lv_color_hsv_to_rgb(uint16_t h, uint8_t s, uint8_t v)
 }
 
 /**
- * Convert an RGB color to HSV
- * @param r red
- * @param g green
- * @param b blue
- * @return the given RGB color n HSV
+ * Convert a 32-bit RGB color to HSV
+ * @param r8 8-bit red
+ * @param g8 8-bit green
+ * @param b8 8-bit blue
+ * @return the given RGB color in HSV
  */
-lv_color_hsv_t lv_color_rgb_to_hsv(uint8_t r, uint8_t g, uint8_t b)
+lv_color_hsv_t lv_color_rgb_to_hsv(uint8_t r8, uint8_t g8, uint8_t b8)
 {
+    uint16_t r = ((uint32_t)r8 << 10) / 255;
+    uint16_t g = ((uint32_t)g8 << 10) / 255;
+    uint16_t b = ((uint32_t)b8 << 10) / 255;
+
+    uint16_t rgbMin = r < g ? (r < b ? r : b) : (g < b ? g : b);
+    uint16_t rgbMax = r > g ? (r > b ? r : b) : (g > b ? g : b);
+
     lv_color_hsv_t hsv;
-    uint8_t rgbMin, rgbMax;
 
-    rgbMin = r < g ? (r < b ? r : b) : (g < b ? g : b);
-    rgbMax = r > g ? (r > b ? r : b) : (g > b ? g : b);
+    // https://en.wikipedia.org/wiki/HSL_and_HSV#Lightness
+    hsv.v = (100 * rgbMax) >> 10;
 
-    hsv.v = rgbMax;
-    if(hsv.v == 0) {
+    int32_t delta = rgbMax - rgbMin;
+    if (LV_MATH_ABS(delta) < 3) {
         hsv.h = 0;
         hsv.s = 0;
         return hsv;
     }
 
-    hsv.s = 255 * (long)(rgbMax - rgbMin) / hsv.v;
-    if(hsv.s == 0) {
+    // https://en.wikipedia.org/wiki/HSL_and_HSV#Saturation
+    hsv.s = 100 * delta / rgbMax;
+    if(hsv.s < 3) {
         hsv.h = 0;
         return hsv;
     }
 
+    // https://en.wikipedia.org/wiki/HSL_and_HSV#Hue_and_chroma
+    int32_t h;
     if(rgbMax == r)
-        hsv.h = 0 + 43 * (g - b) / (rgbMax - rgbMin);
+        h = (((g - b) << 10) / delta) + (g < b ? (6 << 10) : 0); // between yellow & magenta
     else if(rgbMax == g)
-        hsv.h = 85 + 43 * (b - r) / (rgbMax - rgbMin);
+        h = (((b - r) << 10) / delta) + (2 << 10); // between cyan & yellow
```

Comment:
Won't it cause an error if `r > b` as @tgillbe pointed out?

Reply:
I don't believe so. That would just be a negative number, and the `h < 0` check later bumps it back to positive.